client/service/src/client/client.rs

```diff
@@ -831,58 +847,73 @@ where
 		operation.op.mark_finalized(BlockId::Hash(block), justification)?;
 
 		if notify {
-			// sometimes when syncing, tons of blocks can be finalized at once.
-			// we'll send notifications spuriously in that case.
-			const MAX_TO_NOTIFY: usize = 256;
-			let enacted = route_from_finalized.enacted();
-			let start = enacted.len() - std::cmp::min(enacted.len(), MAX_TO_NOTIFY);
-			for finalized in &enacted[start..] {
-				operation.notify_finalized.push(finalized.hash);
+			let finalized =
+				route_from_finalized.enacted().iter().map(|elem| elem.hash).collect::<Vec<_>>();
+
+			// Prevent inclusion of stale heads that were already included in previous calls.
+			let last_finalized_number = self
+				.backend
+				.blockchain()
+				.number(last_finalized)?
+				.expect("Finalized block expected to be onchain. qed");
+			let mut stale_heads = Vec::new();
+			for head in self.backend.blockchain().leaves()? {
+				let route_from_finalized =
+					sp_blockchain::tree_route(self.backend.blockchain(), block, head)?;
+				let retracted = route_from_finalized.retracted();
+				let pivot = route_from_finalized.common_block();
+				if !(retracted.is_empty() || last_finalized_number > pivot.number) {
```

Comment:
Can you explain the second condition? If nothing is retracted then the common block should be `block`, in which case the `last_finalized_number` is never higher. Is my thinking correct here?

Reply:
I spotted that `backend.blockchain().leaves()` may return heads that were already staled in previous finalization/s and thus already included in prev notifications.

Motivation: the `Backend` implementation for `BlockchainDb` stores the leaves in memory [here](https://github.com/paritytech/substrate/blob/4cd2cc958f5a078d87e4eae877167c40cf2ebf9e/client/db/src/lib.rs#L449-L456) and being stale is not a sufficient condition to immediately remove the head from the list, see [here](https://github.com/paritytech/substrate/blob/4cd2cc958f5a078d87e4eae877167c40cf2ebf9e/client/api/src/leaves.rs#L131-L134) why.

Said that, we want to avoid returning stale heads that were already included in previous finalizations.

Lets split the condition in two for easier analysis.
    `!retracted.is_empty() && !(last_finalized_number > pivot.number)`

To be added to stale heads list, first of all `retracted` from current finalized block to the head should not be empty.

But this last condition is not sufficient to avoid including heads that were already included in previous finalization calls.

An efficient way to check this condition is by controlling that the block height of the previously finalized block is not greater than the pivot (between new finalized and head we are checking), in other words:
- if last_finalized_number <= pivot.number then this is a new stale head
- else the stale head was already included by some previous finalization.

Reply:
I think adding a reduced version of this as a comment would be helpful for future readers, particularly the fact that stale heads might not be immediately pruned from `backend.leaves()`.

Reply:
Done